components/datetime/src/pattern/parser.rs

```diff
@@ -1,110 +1,430 @@
+    ) -> Result<bool, Error> {
+        if ch == '\'' {
+            match (&mut self.state, chars.peek() == Some(&'\'')) {
+                (
+                    Segment::Literal {
+                        ref mut literal, ..
+                    },
+                    true,
+                ) => {
+                    literal.push('\'');
+                    chars.next();
+                }
+                (Segment::Literal { ref mut quoted, .. }, false) => {
+                    *quoted = !*quoted;
+                }
+                (Segment::Symbol { symbol, length }, true) => {
+                    result.push((*symbol, *length).try_into()?);
+                    self.state = Segment::Literal {
+                        literal: String::from(ch),
+                        quoted: false,
+                    };
+                    chars.next();
+                },
+                (Segment::Symbol { symbol, length }, false) => {
+                    result.push((*symbol, *length).try_into()?);
+                    self.state = Segment::Literal {
+                        literal: String::new(),
+                        quoted: true,
+                    };
+                },
+            }
+            Ok(true)
+        } else if let Segment::Literal {
+            ref mut literal,
+            quoted: true,
+        } = self.state
+        {
+            literal.push(ch);
+            Ok(true)
+        } else {
+            Ok(false)
         }
-        Ok(())
     }
 
-    fn collect_symbol(&mut self, symbol: FieldSymbol, idx: usize) -> Result<(), Error> {
-        let field: Field = (symbol, idx - self.state.start_idx).try_into()?;
-        self.items.push(field.into());
-        Ok(())
-    }
-
-    fn collect_literal(&mut self, idx: usize) {
-        if idx > self.state.start_idx {
-            let slice = &self.source[self.state.start_idx..idx];
-            let item = PatternItem::Literal(slice.to_string());
-            self.items.push(item);
+    fn collect_segment(state: Segment, result: &mut Vec<PatternItem>) -> Result<(), Error> {
+        match state {
+            Segment::Symbol { symbol, length } => {
+                result.push((symbol, length).try_into()?);
+            }
+            Segment::Literal { quoted, .. } if quoted => {
+                return Err(Error::UnclosedLiteral);
+            }
+            Segment::Literal { literal, .. } => {
+                if !literal.is_empty() {
+                    result.push(literal.into());
+                }
+            }
         }
+        Ok(())
     }
 
     pub fn parse(mut self) -> Result<Vec<PatternItem>, Error> {
-        for (idx, b) in self.source.bytes().enumerate() {
-            if let Segment::Symbol { byte, .. } = self.state.segment {
-                if b == byte {
-                    continue;
-                }
-            }
+        let mut chars = self.source.chars().peekable();
+        let mut result = vec![];
```

Comment:
Optional: you are keeping some ephemeral state in the class (self.state) and some in local variables (chars, result).  It would result in fewer arguments passed to helper functions if you kept everything in the class.

Reply:
That unfortunately is non-trivial because of how Rust understands partial borrows at the moment. If I move `result `to be a field on `Parser` then I struggle to avoid cloning because I take one field out of it, while passing a mutable reference to a callsite.

I'll keep it as is and as Rust improves and we work on 0.2 I may revisit for cleanups.